README.md
V2V测试联调指南
=====================
源码调测v2v需要搭建如下环境。
在开发机（develop windows）上安装好pycharm社区版。
后续所有命令都需要运行在python 3.6.8及以上的虚拟环境下。
通过pip安装虚拟环境（virtualenv）不赘述，主要过程是先要yum或apt-get安装python3.x的版本。
用安装好的python3.x版本的对应pip3.x来安装一个virtualenv包，再用python运行virtualenv包创建虚拟环境。
最后在虚拟环境的bin目录，source activate，激活虚拟环境。之后在这个虚拟环境下pip安装各种包即可。

![img](docs/images/development_environment.drawio.svg)=98722

快速开始
--------
1.下载代码到本地，安装requirments.txt的包。

```
(v2v) bda@c3daace9a16d:~/work$ git clone https://gitee.com/iovVis/v2v.git v2v
(v2v) bda@c3daace9a16d:~/work$ cd v2v
(v2v) bda@c3daace9a16d:~/work$ pip install -r requirments.txt
(v2v) bda@c3daace9a16d:~/work$ export PYTHONPATH=$PYTHONPATH:/home/bda/work/v2v/src/
(v2v) bda@c3daace9a16d:~/work$ cd tests
```
如果是windows的命令行环境，则通过如下方式设置环境变量PYTHONPATH。
```
(venv) E:\_proj\odoo-14.0.post20201231\v2v\tests>set PYTHONPATH=%PYTHONPATH%;..\src
```
如果是windows的powershell环境，则通过如下方式设置环境变量PYTHONPATH。
```
(venv) E:\_proj\odoo-14.0.post20201231\v2v\tests>$Env:PYTHONPATH += ";..\src"
```
2.在命令行终端界面上，进行流媒体功能测试。

2.1 启动easydarwin服务器，在终端用ffmpeg推流，ffplay测试播放。
easyDarwin服务器在vs目录下的EasyDarwin-windows-8.1.0-1901141151下的EasyDarwin.exe，双击执行。
ffmpeg推流工具和ffplay播放工具在vs目录下的ffmpeg-win64-gpl-vulkan下的bin目录，命令行运行如下。
```
$ ffmpeg -re -stream_loop -1 -i ./plc.mp4 -rtsp_transport tcp -vcodec libx264 -f rtsp rtsp://127.0.0.1:7554/plc
$ ffmpeg -re -stream_loop -1 -i ./panel.mp4 -rtsp_transport tcp -vcodec libx264 -f rtsp rtsp://127.0.0.1:7554/panel
$ ffmpeg -re -stream_loop -1 -i ./person.mp4 -rtsp_transport tcp -vcodec libx264 -f rtsp rtsp://127.0.0.1:7554/person

$ ffplay -rtsp_transport tcp rtsp://192.168.101.19:7554/plc
```
  【注意】easyDarwin为服务器时在easydarwin.ini中设置一下rtsp的端口为7554，要避免与已有端口冲突。

2.2 启动仿真接口。
```
(v2v) bda@c3daace9a16d:~/work/tests$ cd ../src/mock
(v2v) bda@c3daace9a16d:~/work/v2v/src/mock$ python exif.py
```
  【注意】启动仿真接口前要确定，设置对应的流媒体路径与上面的推流地址匹配，在exif.py文件stream_info函数内，大致166等行位置json的url键对应的值。

2.3 在代码的tests路径下启动v2v服务。
```
(v2v) bda@c3daace9a16d:~/work/tests$ python test_main.py
```

3.缺省通过浏览器访问所有的rest接口。
主程序的REST接口
https://127.0.0.1:7080/docs
仿真程序的REST接口
https://127.0.0.1:7180/docs

4.缺省的配置界面。
https://127.0.0.1:7080/ui/v2vlabel/editors/gallery.html

v2v运行的配置文件
--------
v2v程序的正常运行依赖如下一些配置文件。
1.tests目录下的v2v.cfg文件——主配置文件，确定要识别的流媒体源及预置点配置、mqtt服务器地址等、本地启动的微服务端口、视频调度服务器地址等。
这些信息在程序运行后，会由CDM下发配置进行更新。
2.tests目录下的baseconfig.cfg文件——配置jaeger服务器地址/端口等监控需要的信息。
3.tests目录下的logging.conf文件——配置日志文件记录方式。
4.src/mock目录下的cert.cer，cert.key文件——https服务依赖的证书和密钥文件。

v2v运行的监控
--------
1.v2v支持prometheus和jaeger指标监测，需要提前安装Prometheus和Jaeger，命令如下。
```
docker run \
    -p 9090:9090 \
    -v /path/to/prometheus.yml:/etc/prometheus/prometheus.yml \
    prom/prometheus

docker run -d --name jaeger \
  -e COLLECTOR_ZIPKIN_HOST_PORT=:9411 \
  -p 5775:5775/udp \
  -p 6831:6831/udp \
  -p 6832:6832/udp \
  -p 5778:5778 \
  -p 16686:16686 \
  -p 14250:14250 \
  -p 14268:14268 \
  -p 14269:14269 \
  -p 9411:9411 \
  jaegertracing/all-in-one:1.29
```
  【注意】：promethues安装挂出到宿主机的配置文件prometheus.yml参考如下。
```
# my global config
global:
  scrape_interval:     15s # Set the scrape interval to every 15 seconds. Default is every 1 minute.
  evaluation_interval: 15s # Evaluate rules every 15 seconds. The default is every 1 minute.
  # scrape_timeout is set to the global default (10s).

# Alertmanager configuration
alerting:
  alertmanagers:
  - static_configs:
    - targets:
      # - alertmanager:9093

# Load rules once and periodically evaluate them according to the global 'evaluation_interval'.
rule_files:
  # - "first_rules.yml"
  # - "second_rules.yml"

# A scrape configuration containing exactly one endpoint to scrape:
# Here it's Prometheus itself.
scrape_configs:
  # The job name is added as a label `job=<job_name>` to any timeseries scraped from this config.
  - job_name: 'prometheus'

    # metrics_path defaults to '/metrics'
    # scheme defaults to 'http'.

    static_configs:
    - targets: ['localhost:9090']
   
  - job_name: 'v2v'
    scrape_interval: 1s
    metrics_path: /metrics
    scheme: https
    tls_config:
        insecure_skip_verify: true
    static_configs:
    - targets: ['192.168.101.19:7080']
    params:
        v2v: ['rest','rtsp']
```

2.Prometheus的配置。
需要配置Prometheus的prometheus.yml文件（docker通过volume映射在外部，targets的地方ip地址要配置为v2v的运行地址和端口）。
```
scrape_configs:
  # The job name is added as a label `job=<job_name>` to any timeseries scraped from this config.
  - job_name: 'v2v'
    scrape_interval: 1s
    metrics_path: /metrics
    scheme: https
    tls_config:
        insecure_skip_verify: true
    static_configs:
    - targets: ['192.168.101.19:7080']
    params:

docs/history.rst
.. include:: ../HISTORY.rst

2021.12.22
当前配置的主要问题和解决的方案。
配置下发的时候，一种是针对单视频通道的配置下发，一种是针对所有视频通道的全配置文件下发。两种情况都需要支持热更新整个程序，按照新的配置工作。
1）针对单通道的配置下发。
已解决。
2）针对多通道配置的下发。
已解决。
3）遗留问题。
配置信息的热更新。

热更新的方案：
a.不重启进程，而是改rtsp进程的那些控制参数。
b.统计一下rtsp，ai，mqtt都会用哪些控制参数呢：
ai-不需要配置，所有信息是rtsp发来的。

mqtt-需要，涉及到服务器重新连接等操作，破坏了原基类假定的startup，loop，shutdown三步，需要在基类进行改造。
mqtt比较简单，只要重新连接服务器就可以了。
*已经OK。mqtt线程能够自动重启，当收到配置更新的消息时候。
主要的实现方案：在startup中，找主进程要配置信息，要到后把自己关心的配置信息全部设置为成员变量，然后连接mqtt服务器，如果出错，抛出V2V异常。
procworker基类主循环不处理异常，继续抛给基类，基类发现是V2V异常，就重新启动新一轮循环。
这种实现方案有两个地方会触发本进程重启，即：1）收到配置信息更新；2）刚开始启动的时候出现配置错误（因为捕获了所有的线程，所以不会退出。
已经回归测试，能够发送stop的时候，退出流水线。


rtsp-需要。
此前的方案是在创建进程的时候给pipeline分配任务参数，比如url，mqtt的连接地址等。这样在重新启动的时候就比较难处理，
最理想的方式是启动的流水线应该为无任务绑定的。当任务来了以后，流水线自己取领取任务。也就是在创建任务的时候，只是启动了若干流水线，
在startup的时候，大家取获取任务。如果任务分配完，就算了。

1)rtsp采用GrabFrame去取流。
2)rtsp多个进程去要任务，任务列表分配在stop pipe line 和 下发配置的时候被清空。
3)rtsp进程多于实际任务，空闲进程会休息5秒再去要求工作。
4)rtsp进程少于实际任务，...
测试要求，先测正常下发功能，再测正常后断网恢复，再测开始断网后恢复。
1)rtsp进程多于等于新下发任务数，预期：多的子进程休息5秒，重新申请任务，测试通过。（OK）
2)rtsp进程少于新下发任务数，新任务不被执行，但配置更新，重新stop，start整个管道流水线，会按照新的任务数量启动进程执行。（OK）
3)第1/2个测试场景，正常流水线执行态，断网，恢复网络。（OK）
4)第1/2个测试场景，先断网，启动流水线，恢复网络。（OK）

GrabFrame的错误恢复机制被设计为：1）只要open_stream成功，以后read_frame如果不成功，就会再次启动内部线程去重连。这样能避免在上层
处理网络不稳定中断的情况。2）如果刚开始没有配置对或连不成功，则直接返回错误，不做多余的恢复处理。这个错误的恢复交给上层业务，上层调用，
发现不成功，则再重调用就行了。

c.不管是rtsp还是mqtt，如果不重启进程，就需要重写基类的run的逻辑。为了最大限度保持原来逻辑不变，采用抛出异常的方案。
当procworker基类在mainloop中执行main_func的时候，如果收到广播的配置更改的消息，则抛出一个自定义的异常。在基类run中处理。

2021.12.23
1）在rest的presets的路由调用时，进行处理，先通知主进程暂停，然后开始，最后再恢复。

2021.12.24
处理配置文件的热更新问题。


2021.12.25
当前通道配置aoi的时候，会临时取得摄像头的控制权，这会打断识别流水线，之后再释放控制权。
并且要支持多任务之间的互斥。

2021.12.30
配置热更新的工作已经完成。
*特殊处理：在rtsp的startup中，网络下发新配置，在shutdown中要清空任务，否则原来那个进程还会自己跑不停。

目前的逻辑：
基类处理下发配置事件，子类不用管，只要正确startup和shutdown，做好现场清理工作。
如果出现配置错误，首次断网情况，则startup直接抛出异常，基类不会调用main_loop，shutdown。
抛出的异常中，有两个特殊异常：空任务异常表明当前没活干，所以休眠5秒再要任务；配置错误异常，立刻重新进入startup/main_loop/shutdown循环。
其它的异常则，该进程退出。这会导致后面的ai进程，mqtt进程等没有输入，整个流水线后面的进程是无用的，需要发送关闭流水线才能收回进程资源。

2021.12.31
1）jaeger配置；
2）只关闭配置通道对应的视频，并且不再尝试重启，因为下发配置会重启它；
3）多用户互斥，只告知有其它用户也在调用，需要有一个全局对象管理，多刷新几次。

2022.1.5
解决rtsp和ai的pic队列的容量限制，避免ai执行过慢把队列挤爆，同时加入rest api的req_id记录视频帧的timestamp的内容。

2022.1.6
1）jaeger配置（OK）；

2022.1.7
1）分开本身固定不变的配置和下发的配置，并发版（OK）。

2022.1.8
1）只关闭配置通道对应的视频，并且不再尝试重启，因为下发配置会重启它；
2）多用户互斥，只告知有其它用户也在调用，需要有一个全局对象管理，多刷新几次；

src/core/rest.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

# Copyright 2021 The CASICloud Authors. All Rights Reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ==============================================================================
# pylint: disable=invalid-name
# pylint: disable=missing-docstring

"""
=========================
rest module
=========================

Provide web api access points.
"""

# Author: Awen <26896225@qq.com>
# License: Apache Licence 2.0
import logging

import uvicorn
import cv2
import imutils
import io
import base64
import os
import time
import psutil
import xml.etree.ElementTree as xmlET

from prometheus_fastapi_instrumentator.metrics import Info
from prometheus_client import Gauge
from typing import Callable
from prometheus_fastapi_instrumentator import Instrumentator
from matplotlib import pyplot as plt
# from imutils.video import VideoStream
from typing import Optional
from pydantic import BaseModel
from fastapi import FastAPI
from fastapi_utils.tasks import repeat_every
from fastapi.staticfiles import StaticFiles
from fastapi.middleware.cors import CORSMiddleware
from os.path import isfile, join
from os import listdir

from utils import bus, comn, log, GrabFrame, wrapper as wpr
from simplegallery.gallery_init import gallery_create
from simplegallery.gallery_build import gallery_build
# from utils.config import ConfigSet
from core.procworker import ProcWorker
from uvicorn.main import Server

# rtsp url timeoff
OPEN_RTSP_TIMEOFF = 30

# 猴子补丁：在退出本进程的时候，ctrl+c会等待较长时间关闭socket
original_handler = Server.handle_exit


class AppStatus:
    should_exit = False

    @staticmethod
    def handle_exit(*args, **kwargs):
        AppStatus.should_exit = True
        original_handler(*args, **kwargs)


Server.handle_exit = AppStatus.handle_exit


# 添加全局指标监测，如cpu、mem等，不需要进程上下文信息
def cpu_rate() -> Callable[[Info], None]:
    metric = Gauge(
        "cpu_rate",
        "cpu占用率.",
        labelnames=("application", "host")
    )

    def instrumentation(info: Info) -> None:
        cpu = psutil.cpu_percent()
        metric.labels('main', 'localhost').set(cpu)     # 可以根据需要设置更多的label标签和值
        info.request.query_params.getlist('v2v')        # 可以获得url参数，http://127.0.0.1:7080/metrics?v2v=x&v2v=2

    return instrumentation


def mem_rate() -> Callable[[Info], None]:
    metric = Gauge(
        "mem_rate",
        "内存占用率.",
        labelnames=("application",)
    )

    def instrumentation(info: Info) -> None:
        mem = psutil.virtual_memory().percent
        metric.labels('main').set(mem)
        info.request.query_params.getlist('v2v')

    return instrumentation


class RestWorker(ProcWorker):
    def __init__(self, name, in_q=None, out_q=None, dicts=None, **kwargs):
        super().__init__(name, bus.EBUS_TOPIC_REST, dicts, **kwargs)    # 不订阅rtsp、ai、mqtt等线程主题，避免被停等
        self.cached_cvobjs_ = {}     # 缓存opencv的封装对象
        self.in_q_ = in_q
        self.out_q_ = out_q

        self.port_ = None
        self.ssl_keyfile_ = None
        self.ssl_certfile_ = None

        for key, value in dicts.items():
            if key == 'port':
                self.port_ = value
            elif key == 'ssl_keyfile':
                self.ssl_keyfile_ = value
            elif key == 'ssl_certfile':
                self.ssl_certfile_ = value

    def create_app(self) -> FastAPI:

        app_ = FastAPI(
            title="视频图像智能分析软件",
            description="视频图像智能分析软件对外发布的RESTful API接口",
            version="2.2.0", )

        # 支持跨域
        origins = ['*']
        app_.add_middleware(
            CORSMiddleware,
            allow_origins=origins,
            allow_credentials=True,
            allow_methods=['*'],
            allow_headers=['*']
        )

        # 全局变量
        rest_proc_ = self
        # cfg_ = None
        baseurl_of_nvr_samples_ = '/viewport'
        # FIXME:这个地方应该改为从主进程获取配置，主进程是唯一来源，子进程避免直接操作文件系统
        cfgobj = self.call_rpc(bus.CB_GET_CFG, {'cmd': 'get_cfg', 'source': self.name})
        localroot_of_nvr_samples_ = cfgobj['nvr_samples']
        ui_config_tpl_ = f'{cfgobj["ui_config_dir"]}ui.xml'
        ai_url_config_file_ = f'{cfgobj["ui_config_dir"]}diagrameditor.xml'

        # localroot_of_nvr_samples_ = ConfigSet.get_cfg()['nvr_samples']
        # ui_config_tpl_ = f'{ConfigSet.get_cfg()["ui_config_dir"]}ui.xml'
        # ai_url_config_file_ = f'{ConfigSet.get_cfg()["ui_config_dir"]}diagrameditor.xml'

        # EIF3:REST V2V C&M 外部接口-提供UI前端配置V2V需要的截图
        # 本路由为前端ui的路径
        app_.mount('/ui', StaticFiles(directory='../src/ui'), name='ui')
        app_.mount('/static', StaticFiles(directory='../src/swagger_ui_dep/static'), name='static')
        # 本路由为thumbnail预览图片保存位置，该位置下按nvr的deviceid建立文件夹，放置所有base64的采样图片
        app_.mount(baseurl_of_nvr_samples_, StaticFiles(directory=localroot_of_nvr_samples_), name='nvr')

        class Switch(BaseModel):
            cmd: str = 'start'

        @app_.post("/api/v1/v2v/pipeline/")
        async def pipeline(item: Switch):
            """统一关闭或启动rtsp，ai，mqtt子进程"""
            cmds = ['start', 'stop']
            if item.cmd in cmds:
                # rest_proc_.send_cmd(bus.EBUS_TOPIC_MAIN, item.cmd) # noqa
                if item.cmd == 'start':
                    ret = rest_proc_.call_rpc(bus.CB_STARTUP_PPL, {'cmd': item.cmd})  # noqa
                else:
                    ret = rest_proc_.call_rpc(bus.CB_STOP_PPL, {'cmd': item.cmd})  # noqa
            else:
                ret = {'reply': 'unrecognized command.'}
            return ret

        class ViewPorts(BaseModel):
            rtsp_url: str = ''
            device_id: Optional[str] = ''
            channel_id: Optional[str] = ''
            name: Optional[str] = ''
            sample_rate: int = 1
            view_ports: str = ''

        @app_.post("/api/v1/v2v/setup_single_channel")
        async def setup_single_channel(cfg: ViewPorts):
            """C&M之C：设置配置文件，收到该配置文件后，v2v将更新单通道的配置文件"""
            """当前功能是接受一个单通道视频设置给到ai"""
            item = {'version': '1.0.0', 'reply': 'pending.'}
            ret = rest_proc_.call_rpc(bus.CB_SET_CFG, cfg.__dict__)  # noqa 调用主进程函数，传配置给它。
            item['reply'] = ret['reply']
            item['desc'] = ret['desc']
            return item

        class Channels(BaseModel):
            version: str = '1.0.0'
            rtsp_urls: str = '[]'
            mqtt_svrs: str = '[]'
            micro_service: str = '{}'
            nvr_samples: str = ''
            ui_config_dir: str = ''
            media_service: str = ''
            ipc_ptz_delay: int = 3

        @app_.post("/api/v1/v2v/setup_all_channels")
        async def setup_all_channels(cfg: Channels):
            """C&M之C：设置配置文件，收到该配置文件后，v2v将更新整个v2v的配置文件"""
            """当前功能是接受整个配置设置给到ai"""
            item = {'version': '1.0.0', 'reply': 'pending.'}
            rest_proc_.call_rpc(bus.CB_SET_CFG, cfg.__dict__)  # noqa 调用主进程函数，传配置给它。
            item['reply'] = True
            return item

        class AiURL(BaseModel):
            person: str = 'https://127.0.0.1:7180/api/v1/ai/person'
            plc: str = 'https://127.0.0.1:7180/api/v1/ai/plc'
            ocr: str = 'https://127.0.0.1:7180/api/v1/ai/panel'
            llm: str = 'https://127.0.0.1:7180/api/v1/ai/panel'
            pointer: str = 'https://127.0.0.1:7180/api/v1/ai/panel'
            switch: str = 'https://127.0.0.1:7180/api/v1/ai/panel'
            indicator: str = 'https://127.0.0.1:7180/api/v1/ai/panel'

        @app_.post("/api/v1/v2v/setup_ai_url")
        async def setup_ai_url(cfg: AiURL):
            """C&M之C：设置ai微服务模型的url"""
            """当前功能是修改配置文件"""
            item = {'version': '1.0.0', 'reply': 'pending.'}
            tree = xmlET.parse(ui_config_tpl_)
            root = tree.getroot()

            person = root.findall('./Array/add/person')
            if len(person) == 1:
                person[0].attrib['href'] = cfg.person

            plc = root.findall('./Array/add/PLC')
            if len(plc) == 1:
                plc[0].attrib['href'] = cfg.plc

            ocr = root.findall('./Array/add/OCR')
            if len(ocr) == 1:
                ocr[0].attrib['href'] = cfg.ocr

            llm = root.findall('./Array/add/LLM')
            if len(llm) == 1:
                llm[0].attrib['href'] = cfg.llm

            pointer = root.findall('./Array/add/METER')
            if len(pointer) == 1:
                pointer[0].attrib['href'] = cfg.pointer

            switch = root.findall('./Array/add/SWITCH')
            if len(switch) == 1:
                switch[0].attrib['href'] = cfg.switch

            indicator = root.findall('./Array/add/IDL')
            if len(indicator) == 1:
                indicator[0].attrib['href'] = cfg.indicator

            tree.write(ai_url_config_file_)
            item['reply'] = True
            return item

        @app_.get("/api/v1/v2v/metrics")
        async def provide_metrics(deviceid: str, channelid: str):
            """C&M之M：接受监控端比如promethus的调用，反馈自己是否在线和反馈各种运行时信息"""
            item = {'version': '1.0.0'}
            rest_proc_.log(deviceid)  # noqa
            rest_proc_.log(channelid)  # noqa

            return item

        # EIF5:REST PTZ CNTL 代理视频调度管理软件，返回当前所有摄像头的描述
        @app_.get("/api/v1/ptz/streaminfo")
        async def stream_info():
            """获取所有的视频通道列表"""
            item = {'version': '1.0.0', 'reply': 'pending.'}
            cfg = rest_proc_.call_rpc(bus.CB_GET_CFG, {'cmd': 'get_cfg', 'source': rest_proc_.name})
            comn.set_common_cfg(cfg)    # 只是设置全局变量，便于配置的热更新（视频调度管理软件的摄像头停留时间和视频调度服务器地址）
            streams = comn.get_urls()
            item['streams'] = streams
            item['reply'] = True
            return item

        @app_.get("/api/v1/v2v/presets/{deviceid}/{channelid}")
        async def get_presets(deviceid: str, channelid: str, refresh: bool = False):
            """获取该视频通道所有预置点，然后逐个预置点取图，保存为base64，加入流断处理，加入流缓存以及互斥锁保护全局缓存流"""
            item = {'version': '1.0.0', 'reply': 'pending.', 'rtsp_url': None}
            try:
                # 设置全局变量，便于热更新配置（视频调度管理软件的摄像头停留时间和视频调度服务器地址在comn中发挥作用。
                cfg = rest_proc_.call_rpc(bus.CB_GET_CFG, {'cmd': 'get_cfg', 'source': rest_proc_.name})
                comn.set_common_cfg(cfg)
                target = f'{localroot_of_nvr_samples_}{deviceid}/'  # 不支持nvrsamples的热更新，因为启动程序需要mount本地目录
                if refresh:
                    # 如果是刷新，需要从nvr取图片保存到本地目录(nvr_samples目录下按设备号创建目录)。
                    url = comn.get_url(deviceid, channelid)
                    presets = None
                    if url:  # url都得不到，就不需要得presets了
                        presets = comn.get_presets(deviceid, channelid)
                    if url and presets:
                        # 通知主调度，暂停pipeline流水线对本摄像头的识别操作，让rtsp流水线rtsp进程sleep多少时间计算得出。
                        rest_p, rtsp_p = self.calculate_delay_time(cfg, url)
                        pipeline_cmd = {'cmd': 'pause', 'deviceid': deviceid, 'channelid': channelid,
                                        'timeout': rtsp_p}
                        isok = rest_proc_.call_rpc(bus.CB_PAUSE_RESUME_PIPE, pipeline_cmd)  # noqa 调用主进程函数，传配置给它。
                        if not isok['reply']:
                            raise RuntimeError(f'Cannot get the control of IPC: {deviceid}, {channelid}.')
                        # rest 需要等待一定时间，让rtsp进程停下来。
                        time.sleep(rest_p)

                        if url in rest_proc_.cached_cvobjs_:
                            # 如果缓存过，就直接用缓存的流。
                            cvobj = rest_proc_.cached_cvobjs_[url]
                        else:
                            # 如果这个url没配置过，则打开一个新的对象来取流。
                            cvobj = GrabFrame.GrabFrame()
                            opened = cvobj.open_stream(url, GrabFrame.GrabFrame.OPEN_RTSP_TIMEOFF)
                            if opened:
                                rest_proc_.cached_cvobjs_[url] = cvobj
                            else:
                                raise cv2.error(f'Failed to open {url}.')
                        # 产生系列预置点图片。
                        item['rtsp_url'] = url  # 前端需要了解是否有合法url，才方便下发配置的时候填入正确的配置值
                        for prs in presets:
                            ret = comn.run_to_viewpoints(deviceid, channelid, prs['presetid'])
                            if ret:
                                # mutex_.acquire()  # 防止流并发访问错误
                                frame = cvobj.read_frame()
                                if frame is None:
                                    cvobj.stop_stream()
                                    rest_proc_.cached_cvobjs_.pop(url, None)    # 如果没有读出数据，清空缓存
                                    raise cv2.error(f'Got null frame from cv2.')
                                # mutex_.release()
                                # height, width, channels = frame.shape
                                # 保存原始图像
                                filename = f'{target}{prs["presetid"]}.png'
                                os.makedirs(os.path.dirname(filename), exist_ok=True)
                                cv2.imwrite(filename, frame)
                                # 保存缩略图，1920x1080的长宽缩小20倍
                                frame = imutils.resize(frame, width=96, height=54)
                                buf = io.BytesIO()
                                plt.imsave(buf, frame, format='png')
                                image_data = buf.getvalue()
                                image_data = base64.b64encode(image_data)
                                outfile = open(f'{target}{prs["presetid"]}', 'wb')
                                outfile.write(image_data)
                                outfile.close()
                    else:
                        # 不是合法的设备号
                        errmsg = f'摄像头{deviceid}-{channelid}查询流地址或预置点失败。'
                        item['reply'] = errmsg
                        raise ValueError(errmsg)
                    # # 通知主调度，恢复pipeline流水线对本摄像头的识别操作
                    # pipeline_cmd['cmd'] = 'resume'
                    # isok = rest_proc_.call_rpc(bus.CB_PAUSE_RESUME_PIPE, pipeline_cmd)  # noqa 调用主进程函数，传配置给它。
                    # if not isok['reply']:
                    #     raise RuntimeError(f'Cannot get the control of IPC: {deviceid}, {channelid}.')
                    #
                    # 不需要通知恢复，因为rest自己先停，等rtsp，然后rtsp停，再等rest，rest操作完，rtsp自动恢复（sleep超时）。
                onlyfiles = [f'{baseurl_of_nvr_samples_}/{deviceid}/{f}'
                             for f in listdir(target) if isfile(join(target, f)) and ('.png' not in f)]
                # 返回视频的原始分辨率，便于在前端界面了解和载入
                # 有问题，如果没有初始化流（fresh=False）,为了获取png文件尺寸，选第一个文件来查询其尺寸
                pngfiles = [fn for fn in listdir(target) if isfile(join(target, fn)) and ('.png' in fn)]
                if len(pngfiles) > 0:
                    item['reply'] = True
                    item['presets'] = onlyfiles
                    pngfile = f'{target}{pngfiles[0]}'
                    item['width'], item['height'] = wpr.get_picture_size(pngfile)
                else:
                    item['reply'] = False
                    item['desc'] = '没有生成可标注的图片。'
            except FileNotFoundError as fs:
                rest_proc_.log(f'{fs}')
            except ValueError as err:
                rest_proc_.log(f'[{__file__}]{err}', level=log.LOG_LVL_ERRO)  # noqa
            except RuntimeError as err:
                rest_proc_.log(f'[{__file__}]{err}', level=log.LOG_LVL_ERRO)  # noqa
            except cv2.error as cve:
                rest_proc_.log(f'[{__file__}]{cve}', level=log.LOG_LVL_ERRO)
            else:
                rest_proc_.log(f'Preset pictures done.')
            finally:
                return item

        class Directory(BaseModel):
            datedir: str = '2022-01-26'

        @app_.post("/api/v1/v2v/pixgallery")
        async def pixgallery(item: Directory):
            """把参数指定日期得视频识别结果打包为可访问的web服务，返回url地址"""
            ret = {'version': '1.0.0', 'reply': False}
            imagedir = f'{localroot_of_nvr_samples_}airesults/{item.datedir}'
            res = gallery_create(imagedir)
            if res:
                res = gallery_build(imagedir)
                if res:
                    visit_url = f'{baseurl_of_nvr_samples_}/airesults/{item.datedir}/public/index.html'
                    ret = {'version': '1.0.0', 'reply': res, 'url': visit_url}
            return ret

        @app_.on_event("startup")
        @repeat_every(seconds=1, wait_first=True)
        def periodic():
            """周期性任务，用于读取系统状态和实现探针程序数据来源的提取，并在空闲时刻消费视频流"""
            # 消耗掉多余的帧
            # if current_video_stream_['url']:
            #     cap = current_video_stream_['videostream']
            #     fps = cap.get(cv2.cv2.CAP_PROP_FPS) + 5  # noqa 消耗完后 grab会返回非真，退出循环。
            #     run = True
            #     while fps > 0 and run:
            #         mutex_.acquire()
            #         run = cap.grab()  # noqa
            #         mutex_.release()
            #         if not run:
            #             rest_proc_.log('Catch up the nvr play speed.')  # noqa
            #         fps -= 1
            pass

        @app_.on_event("shutdown")
        def shutdown():
            """关闭事件"""
            pass

        return app_

    def calculate_delay_time(self, cfgobj, url):
        """
        根据配置信息，计算某设备号，通道号对应视频要等多少时间。
        返回的等待时间有两个值，一个是rest，一个是rtsp。
        1.设想rest发广播消息，rtsp因为会在mainloop中过预置点，这个时间段A一直无法收到消息，需要rest等待。
        2.rest等待A时间后，获取摄像头控制权。
        3.rtsp在0~A时间段都可能收到广播，收到后等待rest操作广播的时间，重新运行。
        如果rtsp不存在，则rest_sleep_time = 0
        如果rtsp存在，
            则按照配置信息中：（seconds x 预置点个数）计时为需要停留的A时间。
            则按照配置信息中：（ipc_ptz_delay*预置点个数）计时为停留的B时间。
        """
        rest_sleep_time = 0
        rtsp_sleep_time = 0
        try:
            ret = self.call_rpc(bus.CB_GET_METRICS, {'cmd': 'get_metrics', 'desc': 'rest api is called.'})
            tasks = ret['result_tasks']
            ipc_ptz_delay = cfgobj['ipc_ptz_delay']
            if url in tasks.keys():
                for urlchannel in cfgobj['rtsp_urls']:
                    if urlchannel['rtsp_url'] == url:
                        presets_size = len(urlchannel['view_ports'])
                        # 多个viewports中第1个预置点的第1个aoi的
                        seconds = list(urlchannel['view_ports'][0].items())[0][1][0]['seconds']
                        rest_sleep_time = seconds * presets_size    # A时间
                        # 如果rtsp刚好rest发送就收到广播，它还是要等一个A时间
                        rtsp_sleep_time = ipc_ptz_delay * presets_size + rest_sleep_time
                        break
        except (KeyError, Exception) as err:
            self.log(f'[{__file__}]{err}', level=log.LOG_LVL_ERRO)
        finally:
            return rest_sleep_time, rtsp_sleep_time

    def up_time(self) -> Callable[[Info], None]:
        metric = Gauge(
            "up_time",
            "开机持续运行时间.",
            labelnames=("application",)
        )
        sts_ = int(time.time())  # 秒为单位
        rest_ = self

        def instrumentation(info: Info) -> None:    # noqa
            # rest_.log(f'Promethues scrape request: {info.request.url}')
            # 主进程的运行时间累计
            nonlocal sts_
            current = int(time.time())
            delta = current - sts_
            metric.labels('main').set(delta)

            # 收集子进程监测数据(数据结构是{'result':{'AIxx':{'up': 1.1, 'xx': 2}, ... }})
            proc_metrics = rest_.call_rpc(bus.CB_GET_METRICS, {'cmd': 'get_metrics', 'desc': 'rest api is called.'})
            res = proc_metrics['result_metrics']
            for key in res.keys():
                item = res[key]
                if 'up' in item.keys():
                    metric.labels(key).set(item['up'])

        return instrumentation

    def run(self):
        localapp = self.create_app()
        # 只要有rest请求，就会触发添加的这几个统计函数，向主进程请求监测数据。
        instrumentator = Instrumentator(
            # excluded_handlers=[".*admin.*", "/metrics"],
        )
        instrumentator.add(self.up_time())
        instrumentator.add(cpu_rate())
        instrumentator.add(mem_rate())
        instrumentator.instrument(localapp).expose(localapp)

        log_config = uvicorn.config.LOGGING_CONFIG
        log_config["formatters"]["default"]["fmt"] = log.get_v2v_logger_formatter()
        log_config["formatters"]["access"]["fmt"] = log.get_v2v_logger_formatter()
        log_config["loggers"]['uvicorn.error'].update({"propagate": False, "handlers": ["default"]})
        uvicorn.run(localapp,  # noqa 标准用法
                    host="0.0.0.0",
                    port=self.port_,
                    ssl_keyfile=self.ssl_keyfile_,
                    ssl_certfile=self.ssl_certfile_,
                    log_level=logging.WARN,
                    log_config=log_config
                    )
